rosbag2_cpp/src/rosbag2_cpp/cache/cache_consumer.cpp

```diff
@@ -0,0 +1,85 @@
+// Copyright 2020, Robotec.ai sp. z o.o.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <memory>
+
+#include "rosbag2_cpp/cache/cache_consumer.hpp"
+#include "rosbag2_cpp/logging.hpp"
+
+namespace rosbag2_cpp
+{
+namespace cache
+{
+
+CacheConsumer::CacheConsumer(
+  std::shared_ptr<MessageCache> message_cache,
+  consume_callback_function_t consume_callback)
+: message_cache_(message_cache),
+  consume_callback_(consume_callback)
+{
+  consumer_thread_ = std::thread(&CacheConsumer::exec_consuming, this);
+}
+
+CacheConsumer::~CacheConsumer()
+{
+  close();
+}
+
+void CacheConsumer::close()
+{
+  ROSBAG2_CPP_LOG_INFO_STREAM(
+    "Writing remaining messages from cache to the bag. It may take a while");
+  is_stop_issued_ = true;
+
+  message_cache_->allow_swap();
+
+  if (consumer_thread_.joinable()) {
+    consumer_thread_.join();
+  }
+}
+
+void CacheConsumer::change_consume_callback(
+  CacheConsumer::consume_callback_function_t consume_callback)
+{
+  std::lock_guard<std::mutex> consumer_lock(consumer_mutex_);
+  consume_callback_ = consume_callback;
```

Comment:
If I understand this correctly, this would technically mean that a single buffer instance can be executed from two different callbacks, is that right? 
Wouldn't it make sense to indicate that the consumer thread should be finished its current execution, so that the new callback is getting executed on an empty buffer?

Reply:
As far as I understand:
Only one callback can be executed by the consumer thread at one time (since only consumer thread executes callbacks).

What could (probably very rarely) happen is that the callback is changed after the buffer is acquired from the cache, but before it is passed to the "old" callback as would happen normally. 

This would cause part of cache to already go into the new bag, which would mean that:
- the former bag could turn out a bit smaller than suggested split size
- 

A bad alternative is to implement a locking mechanism - but change_consume_callback would block the writer.
A much more reasonable alternative is to set the next callback as a member variable and synchronize that - only switching in exec_consuming between loop iterations.

Shall I implement that?

Reply:
> actually, I believe it would be enough to use copy assignment in the consumer loop so that the callback used within is consistent.

That should do from what I understand.